### PR TITLE
Enable Connection creation from Vault parameters

### PR DESCRIPTION
**Pull request by davlum**
Currently using the Vault secrets backends requires that users store
the secrets in connection URI format:
https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#connection-uri-format

Unfortunately the connection URI format is not capable of expressing
all values of the Connection class. In particular the Connection
class allows for arbitrary string values for the  `extra` parameter,
while the URI format requires that this parameter be unnested JSON
so that it can serialize into query parameters.

```
>>> Connection(conn_id='id', conn_type='http', extra='foobar').get_uri()
[2021-03-25 13:31:07,535] {connection.py:337} ERROR - Expecting value: line 1 column 1 (char 0)
Traceback (most recent call last):
  File "/Users/da.lum/code/python/airflow/airflow/models/connection.py", line 335, in extra_dejson
    obj = json.loads(self.extra)
  File "/nix/store/8kzdflq0v06fq0mh9m2fd73gnyqp57xr-python3-3.7.3/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/nix/store/8kzdflq0v06fq0mh9m2fd73gnyqp57xr-python3-3.7.3/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/nix/store/8kzdflq0v06fq0mh9m2fd73gnyqp57xr-python3-3.7.3/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
[2021-03-25 13:31:07,535] {connection.py:338} ERROR - Failed parsing the json for conn_id id
'http://'
```

As shown, the `extra` data is missing from the return value `http://`.
Although there is an error logged, this does not help users who were
previously able to store other data.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by dstandish** [comment]
> Unfortunately the connection URI format is not capable of expressing
all values of the Connection class. In particular the Connection
class allows for arbitrary string values for the extra parameter,
while the URI format requires that this parameter be unnested JSON
so that it can serialize into query parameters.

That's not true.

We have a convenience method `get_uri` you can use to produce the URI.

Try creating your connection as desired with init params including your json `extra`.  Then call `get_uri` on the connection.  Use this URI.

**Comment by nathadfield** [comment]
> > Unfortunately the connection URI format is not capable of expressing
> > all values of the Connection class. In particular the Connection
> > class allows for arbitrary string values for the extra parameter,
> > while the URI format requires that this parameter be unnested JSON
> > so that it can serialize into query parameters.
> 
> That's not true. Or at least I don't _think_ that's true.
> 
> We have a convenience method `get_uri` you can use to produce the URI.
> 
> Try creating your connection as desired with init params including your json `extra`. Then call `get_uri` on the connection.
> 
> Use this URI in vault.
> 
> If you have a json value for extra that can not be serialized to airflow URI please post it here.

@dstandish Even so, might not this be a good feature anyway?

A `conn_uri` is fine but they can be a bit unwieldy and it is not that straightforward to create one correctly.  I would personally also like the ability to store the individual elements of a connection as a secret in Vault/GSM and for Airflow to build the connection at runtime.

**Comment by dstandish** [comment]
There is a benefit inherent in having only one perfectly good way to do things.

Sometimes when you want to do something another way it's best kept in a private repo for your company or posted on a blog.

I would be in favor of only using the airflow conn uri format.  But let's see what others think -- I'm just one voice here so don't be too discouraged.

In the meantime I encourage you to try the uri generation function we have.  There is documentation on generating the uri in the managing connections doc.

**Comment by dstandish** [comment]
I see you added an example where extra is not valid json.  Why do you want to do that?  Why not just use JSON?  E.g. if you put fubar in quotes probably it works

**Comment by nathadfield** [comment]
> There is a benefit inherent in having only one perfectly good way to do things.
> 
> Sometimes when you want to do something another way it's best kept in a private repo for your company or posted on a blog.
> 
> I would be in favor of only using the airflow conn uri format. But let's see what others think -- I'm just one voice here so don't be too discouraged.
> 
> In the meantime I encourage you to try the uri generation function we have. There is documentation on generating the uri in the managing connections doc.

@dstandish Sure, I take your point but it is also nice to have options.

For the record, I have tried the `conn_uri` method.  Its fine but it does require a certain level of knowledge in order to create one.  There are a certain category of users who may only ever interact with the Airflow UI but may also need to update a secret used by a connection.

**Comment by dstandish** [comment]
Yeah you make a good point that it can't store arbitrary json _(I do imagine we could add support for this within the URI format)_.

Based on that assumption, we do have [fairly comprehensive tests](https://github.com/apache/airflow/blob/master/tests/models/test_connection.py#L308-L358) that verify that URI parsing and URI generation produce consistent results.

But it's true we don't test the case where you store an arbitrary value in extra such as `hello` or `fubar`.

Currently it is assumed dict, so like `{'hello': 'fubar'}`

Anyway I'll let others chime in at this point.

**Comment by avieth** [comment]
In the documentation of the `Connection` class, it is stated that `extra` should be encoded JSON.

https://github.com/apache/airflow/blob/6b9b0675c5ece22a1c382ebb9b904fb18f486211/airflow/models/connection.py#L87

However, AFAICT, the constructor will not check this, and will not fail when `extra` is not JSON. But `get_uri` _does_ check this (implicitly, by trying to decode then urlencode), and so we have the very weird case in which we can construct a `Connection` object for which `get_uri` fails, as @davlum has shown. I'd say that's fairly counterintuitive and surprising.

Is there a good reason why `extra` must be encoded JSON? Seems like a nice solution would be to allow for any string there.

**Review by kaxil — COMMENTED** [review]
Can you please update docs: https://github.com/apache/airflow/blob/master/docs/apache-airflow-providers-hashicorp/secrets-backends/hashicorp-vault.rst

**Comment by dstandish** [comment]
@avieth 

>However, AFAICT, the constructor will not check this, and will not fail when extra is not JSON. But get_uri does check this (implicitly, by trying to decode then urlencode), and so we have the very weird case in which we can construct a Connection object for which get_uri fails, as @davlum has shown. I'd say that's fairly counterintuitive and surprising.
> Is there a good reason why extra must be encoded JSON? Seems like a nice solution would be to allow for any string there

OK i was able to add support in get_uri for arbitrary i.e. non-json string in #15100.  behavior is still to return `{}` when you call `extra_dejson` in this case.  However, it can now be serialized to URI and back without loss.

**Comment by mik-laj** [comment]
> Is there a good reason why extra must be encoded JSON? Seems like a nice solution would be to allow for any string there.

We expect the extra field to be a Web UI dictionary. The use of primitive types also causes a problem with extending a given connection ie. it is not future-proff. We also have a design misunderstood here. This field should contain extra **fields**, not additional data.

**Comment by davlum** [comment]
@mik-laj 

>  We expect the extra field to be a dictionary in Web UI.

This must be as of Airflow 2.0, because this isn't the case in 1.10.x. The additional validation should have been added on the `Connection` object to because it is still perfectly valid to add any data to the `extra` column.

**Comment by mik-laj** [comment]
> I would be in favor of only using the airflow conn uri format. But let's see what others think -- I'm just one voice here so don't be too discouraged.

I like the idea of supporting other formats. LocalFilesystemBackend supports already object representation also: 
https://github.com/apache/airflow/blob/a4aee3f1d0c27f7c6010e784611ee943009c7498/airflow/secrets/local_filesystem.py#L188-L223

**Comment by mik-laj** [comment]
>  The additional validation should have been added on the Connection object 

Yes. I think we should add it. Can you help with it?

**Comment by davlum** [comment]
> Can you help with it?

Sure thing, I'll open another PR for that. Curious what the release target will be because it will be a breaking change.

So just to summarize where we're at:
* Support object backend on Vault in this PR?
* Add validation on `extra` column in another PR

**Comment by mik-laj** [comment]
> Curious what the release target will be because it will be a breaking change.

This will be the release manager's decision, but I would consider that this is not a breaking change as the design assumption of this feature was to store objects that have a name and a value. Now we only improve this feature by adding validation. So I think, we should release it in Airflow 2.1.

**Comment by mik-laj** [comment]
> Seems like a nice solution would be to allow for any string there.

Do you have any use cases that you can't solve with JSON? Object representation is more future-proof because you can always add a new key and tag the old one as deprecaated. It allows for smooth updates.

**Comment by dstandish** [comment]
@mik-laj curious what do you think about deprecating the conn uri format and replacing with JSON, or perhaps allowing both globally? We could implement backward compat with try json parse except.

**Comment by mik-laj** [comment]
I think we support both. The specific choice depends on the specific use case. To be able to use the same value as an environment variable and to copy easily to the clipboard, you can use a URI.  for more convenient editing, you can use object representation.

**Comment by dstandish** [comment]
> I think we support both. The specific choice depends on the specific use case. To be able to use the same value as an environment variable and to copy easily to the clipboard, you can use a URI. for more convenient editing, you can use object representation.

No I mean deprecate across airflow including with env vars ( or support both simultaneously )

There's not really anything special about the uri format.  We could store json in env vars

**Comment by mik-laj** [comment]
I see no reason to do this.  Too many users already use URIs.  Besides, for simple connections it is a very good representation.

File: tests/providers/hashicorp/secrets/test_vault.py
Comment (dstandish) [inline tests/providers/hashicorp/secrets/test_vault.py]: probably should include `extra` since it's a case that should be supported

one thought i had is, when storing creds in the "object" style, should we make it so that you can add values here at top level and they'll automatically get added to extra.

so e.g. we could support this:
```
                'data': {
                    'conn_type': 'postgresql',
                    'login': 'airflow',
                    'password': 'airflow',
                    'host': 'host',
                    'port': '5432',
                    'myfield1': 'hello',
                    'myfield2': 'hello again',
                    'schema': 'airflow',
                },
```

and then when bulding the connection object, add these to the extra dict.

it's maybe a little wacky but it would allow for less nesting.  and then when defining the values in vault you wouldn't have to dump to json and store under `extra` as json string.  just a thought 🤷‍♂️

Comment (davlum) [reply tests/providers/hashicorp/secrets/test_vault.py]: I'm going to piggyback off of the validation that happens in `LocalFilesystemBackend` so I don't think that will be possible.

File: airflow/providers/hashicorp/secrets/vault.py
Comment (dstandish) [inline airflow/providers/hashicorp/secrets/vault.py]: i think what makes sense to do here is to promote `_create_connection` to a standard way of parsing creds from json _(and a non-private function)_, so that all such implementations can use it, and so that it is not associated with one particular backend (in this case local files)

i think the right place for this would probably be in connection.py but i could see it going in the root secrets module / package also.

what do you think @mik-laj

File: tests/providers/hashicorp/secrets/test_vault.py
Comment (dstandish) [reply tests/providers/hashicorp/secrets/test_vault.py]: ah i see or more than validation you are reusing the conn generation logic.  i have commented suggesting you promote that  bit of code.

meanwhile, don't you think the mock here should include the extra field?

File: airflow/providers/hashicorp/secrets/vault.py
Comment (mik-laj) [inline airflow/providers/hashicorp/secrets/vault.py]: We should extract the common part to the new package. This is problematic now, so I started discussions on the mailing list. See: https://lists.apache.org/thread.html/r713f180120d0a39b53567812eb5db34f992ec81979818d2175598b71%40%3Cdev.airflow.apache.org%3E

File: airflow/providers/hashicorp/secrets/vault.py
Comment (mik-laj) [reply airflow/providers/hashicorp/secrets/vault.py]: I support this idea, but it can be a bit problematic now.
See:  https://github.com/apache/airflow/pull/15013#discussion_r606649464

File: airflow/providers/hashicorp/secrets/vault.py
Comment (davlum) [reply airflow/providers/hashicorp/secrets/vault.py]: In this particular case I can see this function existing in `airflow.models.connections` as `from_dict() -> Connection`. Transforming a dict to the object is the sort of function that usually exists with the class. I agree with the general case that something needs to be done about common functionality between providers.

Comment (mik-laj) [reply airflow/providers/hashicorp/secrets/vault.py]: You can't add a new feature to the core because we want to be backwards compatible with Airflow 2.0.0.  I opened the discussions to loosen this restriction, because they do not conform to reality. I think we should maintain backward compatibility with the MINOR release, not the MAJOR. 

I invite you to the discussion on the mailing list if you would like to share your thoughts.

Comment (davlum) [reply airflow/providers/hashicorp/secrets/vault.py]: It's an additional function on the Connection class so that retains backwards compatibility no?

Comment (mik-laj) [reply airflow/providers/hashicorp/secrets/vault.py]: Secret Backend won't work with Airflow 2.0.0, so this is a breaking change to providers packages. This doesn't make breaking changes to Airflow 2.0.0, but to providers packages it does.

Comment (dstandish) [reply airflow/providers/hashicorp/secrets/vault.py]: So you are saying that if this vault provider expects a `from_dict` method on `Connection` (say, to be added in 2.1) the provider won't work with 2.0, but we've promised that it will.

should we perhaps then just merge this as is?  or perhaps duplicate this private function into a vault provider utils module with a note of some kind to replace with Connection.from_dict when it is available?

Comment (mik-laj) [reply airflow/providers/hashicorp/secrets/vault.py]: Yes. Exactly.

We should discuss the approach on the mailing list and choose the best solution that everyone accepts.  I personally think that we should mark this package as only supported by Airflow 2.1, because trying to maintain backward compatibility will limit our development possibilities. It should be normal for users that new packages/library versions may require a core version if we add new features. See: https://lists.apache.org/thread.html/r713f180120d0a39b53567812eb5db34f992ec81979818d2175598b71%40%3Cdev.airflow.apache.org%3E

Comment (mik-laj) [reply airflow/providers/hashicorp/secrets/vault.py]: I started discussions on this subject on the mailing list, but I haven't checked it recently. I encourage you to discuss it yourself because then we will be able to work out a solution to this problem faster.

https://lists.apache.org/thread.html/r713f180120d0a39b53567812eb5db34f992ec81979818d2175598b71%40%3Cdev.airflow.apache.org%3E

Comment (mik-laj) [reply airflow/providers/hashicorp/secrets/vault.py]: I have the impression that we have managed to reach a consensus. This change will be in the next provider package after Airflow 2.1 is released. There is a slight problem with the provider packages releases process. We always release providers packages from the main branch, so we cannot merge this PR until we release Airflow 2.1 and we will start releasing providers packages that will be compatible with this version.

In the meantime, we can create a new public method that will create connections from the dictionary. This method will be released in Airflow 2.1.  PR: https://github.com/apache/airflow/pull/15425

I added this PR to [the Airflow 2.1](https://github.com/apache/airflow/milestone/8) milestone.

**Comment by mik-laj** [comment]
> We cannot merge this PR until we release Airflow 2.1 and we will start releasing providers packages that will be compatible with this version.

More info: https://github.com/apache/airflow/pull/15013#discussion_r615969275

**Comment by jhtimmins** [comment]
@mik-laj are we able to merge this now? Or is there something else we're waiting on?

**Comment by ashb** [comment]
@jhtimmins Unfortunately not, as https://github.com/apache/airflow/pull/15425 didn't get finished (or possibly reviewed! Sorry) in time to make it for 2.1

**Comment by kaxil** [comment]
marked it as 2.2 since it is not a bug fix

**Review by ashb — COMMENTED** [review]
@natanweinberger  Could you take a look and review this PR/tell us what changes are needed now your other change has landed?

File: airflow/providers/hashicorp/secrets/vault.py
Comment (natanweinberger) [inline airflow/providers/hashicorp/secrets/vault.py]: This won't return a connection object. I think this should be
```python
if uri:
    return Connection(conn_id, uri=uri)

return Connection(conn_id, **response) 
# or _create_connection, but constructor handles this logic pretty well now without secrets backend
```

Comment (natanweinberger) [reply airflow/providers/hashicorp/secrets/vault.py]: @ashb Based on the changes we made to the constructor for `Connection`, I don't see any reason not to use it here too.

That said, note that this function won't persist the connection to the DB (but neither would `_create_connection`, so no difference really).

Comment (davlum) [reply airflow/providers/hashicorp/secrets/vault.py]: Wait so what are the changes I should make?

Comment (natanweinberger) [reply airflow/providers/hashicorp/secrets/vault.py]: I was raising two suggestions:
- the Connection constructor can replace the `_create_connection` function for this use case
- in one case, this function returns a Connection object; in another it returns a uri. I think it should return a Connection both times?

File: docs/apache-airflow-providers-hashicorp/secrets-backends/hashicorp-vault.rst
Comment (ashb) [inline docs/apache-airflow-providers-hashicorp/secrets-backends/hashicorp-vault.rst]: Do we need to say this again? Maybe it should be moved to a "setup" section in this doc instead

**Comment by ashb** [comment]
(Code on this looks good now! A nice simple change too!)

**Review by ashb — APPROVED** [review]

**Pull request merged**
